nx/lib/nx/defn/grad.ex

```diff
@@ -424,6 +424,50 @@ defmodule Nx.Defn.Grad do
     to_grad(x, g, cache)
   end
 
+  defp grad(:concatenate, [tensors, axis], %{shape: ans_shape} = ans, g, cache) do
+    operand_shapes = Enum.map(tensors, & &1.shape)
+    limit_points = Enum.scan(operand_shapes, 0, fn x, acc -> elem(x, axis) + acc end)
+
+    n_tensors = length(tensors)
+
+    axes = Nx.axes(ans_shape)
+
+    starts =
+      Enum.map(0..(n_tensors - 1), fn row ->
+        Enum.map(axes, fn col ->
+          if col == axis and row > 0 do
+            Enum.fetch!(limit_points, row - 1)
+          else
+            0
+          end
+        end)
+      end)
+
+    # TODO: Use Enum.zip_with/3 on Elixir v1.12
+    limits =
+      ans_shape
+      |> Tuple.to_list()
+      |> List.duplicate(n_tensors)
+      |> Enum.zip(limit_points)
+      |> Enum.map(fn {row, p} -> List.replace_at(row, axis, p) end)
+
+    ones = Nx.map(ans, fn _ -> 1 end)
```

Comment:
@josevalim Upon inspecting the following code from Jax, I found that there was a `1`s array with shape `ans.shape` arising somewhere.

```python
jax.make_jaxpr(jax.grad(lambda x: jnp.sum((jnp.concatenate((jnp.power(x, 2), jnp.power(x, 3)), 0)))))(np.array([[1.0, 2.0], [3.0, 4.0]]))

{{ lambda  ; a.
  let b = integer_pow[ y=2 ] a
      c = integer_pow[ y=1 ] a
      d = mul 2.0 c
      e = integer_pow[ y=3 ] a
      f = integer_pow[ y=2 ] a
      g = mul 3.0 f
      h = concatenate[ dimension=0 ] b e
      _ = reduce_sum[ axes=(0, 1) ] h
      i = broadcast_in_dim[ broadcast_dimensions=(  )
                            shape=(4, 2) ] 1.0
      j = slice[ limit_indices=(2, 2)
                 start_indices=(0, 0)
                 strides=None ] i
      k = slice[ limit_indices=(4, 2)
                 start_indices=(2, 0)
                 strides=None ] i
      l = mul k g
      m = mul j d
      n = add_any l m
  in (n,) }
```

After defining this array as the one we apply the slices, everything is working as expected. I'm not really sure if this is the correct fix, but if it isn't, this might give some insight on what the fix is. Any thoughts?

After we settle on this aspect, I plan on refactoring some of the code because I think that we can build `starts` and `limits` iteratively in a `Enum.reduce(tensors, ...)` call

Reply:
Update: this implementation makes the tests pass, which makes me believe I have gotten something *really* wrong somewhere:

```elixir
defp grad(:concatenate, [tensors, _axis], _ans, g, cache) do
  tensors
  |> Enum.map(fn t ->
    {t, Nx.map(t, fn _ -> 1 end)}
  end)
  |> grad_pairs(g, cache)
end
```

Reply:
After giving this some thought, I think this makes sense...

If we define a 2 tensor concatenation as
```
concat(A, B) = [[eye(shape(A)) zeros(shape(B))], [zeros]] dot concat(A, 0) + [[zeros], [zeros(shape(A)) ones(shape(B))]] dot concat(0, B)
                     = C1 dot A' + C2 dot B'
```

Then we could calculate grad(concat(A,B)) as follows

```
grad(concat(A, B)) = grad(C1) dot A' + C1 dot grad(A') + grad(C2) dot B' + C2 dot grad(B')
                              # assuming grad(C1) = grad(C2) = 0
                              = C1 dot grad(A') + C2 dot grad(B')
                              = concat(grad(A), grad(B))
```

This would mean that `grad(sum(concat([a, b]))` would result in `grad(a) + grad(b)`, which is indeed the actual result we get in the test cases. For example:

```elixir
defn my_grad_power(t, n \\ 2) do
  grad(t, Nx.sum(Nx.power(t, n)))
end

defn concatenate_grad_power(t) do
  grad(
    t,
    Nx.sum(Nx.concatenate([Nx.power(t, 2), Nx.power(t, 3)], axis: 0))
  )
end

test "computes grad for powers of a {2, 2}-tensor" do
  IO.inspect(my_grad_power(Nx.tensor([[1.0, 2.0], [3.0, 4.0]])))
  IO.inspect(my_grad_power(Nx.tensor([[1.0, 2.0], [3.0, 4.0]]), 3))

   assert concatenate_grad_power(Nx.tensor([[1.0, 2.0], [3.0, 4.0]])) ==
      Nx.tensor([[5.0, 16.0], [33.0, 56.0]])
end
```

Passes and prints:
```elixir
#Nx.Tensor<
  f32[2][2]
  [
    [2.0, 4.0],
    [6.0, 8.0]
  ]
>
#Nx.Tensor<
  f32[2][2]
  [
    [3.0, 12.0],
    [27.0, 48.0]
  ]
>
```

I'm not sure if I followed the code correctly, but `grad_pairs` seems to end up either ignoring this value (`x`) calling `Nx.multiply(t, x)`, which would yield `t` if `x == ones(shape(t))`

Any thoughts on this?